File: iosApp/iosApp/p2p/P2PService.swift
//
//  P2PService.swift
//  iosApp
//
//  Created by jimmyt on 6/4/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import CryptoKit
import Foundation
import os
import PromiseKit
import Switrix

/**
 The main Peer-to-Peer Service. It is responsible for listening to new messages in the Commuto Interface Network Matrix room, parsing the messages, and then passing relevant messages to other services as necessary.
 */
class P2PService {
    
    /**
     Initializes a new P2PService object.
     
     - Parameters:
        - errorHandler: An object that adopts the `P2PErrorNotifiable` protocol, to which this will pass errors when they occur.
        - offerService: An object that adopts the `OfferMessageNotifiable` protocol, to which this will pass offer-related messages.
        - swapService: An object that adopts the `SwapMessageNotifiable` protocol, to which this will pass swap-related messages.
        - switrixClient: A `SwitrixClient` instance used to interact with a Matrix Homeserver.
        - keyManagerService: A `KeyManagerService` from which this gets key pairs when attempting to decrypt encrypted messages.
     
     - Returns: A new `P2PService` instance.
     */
    init(
        errorHandler: P2PErrorNotifiable,
        offerService: OfferMessageNotifiable,
        swapService: SwapMessageNotifiable,
        switrixClient: SwitrixClient,
        keyManagerService: KeyManagerService
    ) {
        self.errorHandler = errorHandler
        self.offerService = offerService
        self.swapService = swapService
        self.switrixClient = switrixClient
        self.keyManagerService = keyManagerService
    }
    
    /**
     P2PService's `Logger`.
     */
    private let logger = Logger(subsystem: "xyz.commuto.interfacemobile", category: "P2PService")
    
    /**
     An object to which `P2PService` will pass errors when they occur.
     */
    private let errorHandler: P2PErrorNotifiable
    
    /**
     An object to which `P2PService` will pass offer-related messages when they occur.
     */
    private let offerService: OfferMessageNotifiable
    
    /**
     An object to which `P2PService` will pass swap-related messages when they occur.
     */
    private let swapService: SwapMessageNotifiable
    
    /**
     A `SwitrixClient` instance used to interact with a Matrix Homeserver.
     */
    let switrixClient: SwitrixClient
    
    /**
     A `KeyManagerService` from which this gets key pairs when attempting to decrypt encrypted messages.
     */
    private let keyManagerService: KeyManagerService
    
    /**
     The token at the end of the last batch of non-empty Matrix events that was parsed. (The value specified here is that from the beginning of the Commuto Interface Network testing room.) This should be updated every time a new batch of events is parsed.
     */
    private var lastNonEmptyBatchToken = "t1-2607497254_757284974_11441483_1402797642_1423439559_3319206_507472245_4060289024_0"
    
    /**
     Used to update `lastNonEmptyBatchToken`. Eventually, this function will store `newToken` in persistent storage.
     
     - Parameter newToken: The batch token at the end of the most recently parsed batch of Matrix events, to be set as `lastNonEmptyBatchToken`.
     */
    private func updateLastNonEmptyBatchToken(_ newToken: String) {
        lastNonEmptyBatchToken = newToken
    }
    
    /**
     The thread in which P2PService listens for and parses messages.
     */
    private var listenThread: Thread?
    
    /**
     The timer that runs in `listenThread`'s `RunLoop` and is responsible for repeatedly executing the listening process.
     */
    private var timer: Timer?
    
    /**
     Indicates whether P2PService is currently handling new data as part of the listening process.
     */
    private var isDoingListening = false
    
    /**
     Creates a new  `Thread` to run `listenLoop()`, updates `listenThread` with a reference to the new `Thread`, and starts it.
     */
    func listen() {
        logger.notice("Starting listen loop in new thread")
        listenThread = Thread { [self] in
            listenLoop()
        }
        listenThread!.start()
        logger.notice("Started listen loop in new thread")
    }
    
    /**
     Invalidates the `Timer` resposible for repeatedly starting the listening process and cancels `listenThread`.
     */
    func stopListening() {
        logger.notice("Invalidating listen loop timer and canceling listen loop thread")
        timer?.invalidate()
        listenThread?.cancel()
        logger.notice("Invalidated listen loop timer and canceled listen loop thread")
    }
    
    /**
     Creates a new `Timer` responsible for repeatedly executing the listening process once approximately every second, adds the `Timer` to the current thread's `Runloop`, and then starts said `Runloop`.
     
     Listening Process:
     
     First,  we check the`isDoingListening` flag. If `isDoingListening` is true, then the listening process is currently running, so we don't start it again. If `isDoingListening` is false, then the listening process isn't currently running, so we can safely start it.
     
     Once we have established that we can safely start the listening process, we call `syncPromise()` to get the most recent batch token from the Matrix homeserver. Then we try to get all new events since that most recent batch token, with a limit of one trillion events, by calling `getMessagesPromise()`. We update our last non-empty batch token with the `nextBatchToken` value in the response from the `syncPromise()` call, and then we parse the new events for relavent messages.
     
     Finally, once the listening process is completed, we set `isDoingListening` to `false` to indicate that the listening process is not currently running and can be started again by the `Timer`.
     
     If we encounter an error, we pass it to our `errorHandler`. Additionally, if the error indicates that we have a missing or invalid token, we call `stopListening()` to prevent the listening process from being started again, since the user must perform some action in order to resolve this issue.
     */
    func listenLoop() {
        let timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [self] _ in
            if (!isDoingListening) {
                logger.notice("Beginning iteration of listen loop")
                isDoingListening = true
                Promise<SwitrixSyncResponse> { seal in
                    DispatchQueue.global(qos: .utility).async {
                        self.syncPromise().pipe(to: seal.resolve(_:))
                    }
                }.then(on: DispatchQueue.global(qos: .utility)) { [self] response -> Promise<(SwitrixGetEventsResponse, String)> in
                    self.logger.notice("Synced with Matrix homeserver, got nextBatchToken: \(response.nextBatchToken)")
                    return getMessagesPromise(from: self.lastNonEmptyBatchToken, limit: 1_000_000_000_000).map { ($0, response.nextBatchToken)  }
                }.done(on: DispatchQueue.global(qos: .utility)) { [self] response, newToken in
                    self.logger.notice("Got new messages from Matrix homeserver from lastNonEmptyBatchToken: \(self.lastNonEmptyBatchToken)")
                    try parseEvents(response.chunk)
                    self.logger.notice("Finished parsing new events from chunk of size \(response.chunk.count)")
                    updateLastNonEmptyBatchToken(newToken)
                    self.logger.notice("Updated lastNonEmptyBatchToken with new token: \(newToken)")
                }.catch(on: DispatchQueue.global(qos: .utility)) { [self] error in
                    self.logger.error("Got an error during listen loop, calling error handler. Error: \(error.localizedDescription)")
                    errorHandler.handleP2PError(error)
                    if (error as NSError).domain == "NSURLErrorDomain" {
                        self.logger.error("Detected error with internet connection; stopping listen loop.")
                        // There is a problem with the internet connection, so there is no point in continuing to listen for new events.
                        stopListening()
                    } else if let switrixError = error as? SwitrixError {
                        switch switrixError {
                        case .unknownToken:
                            self.logger.error("Got unknown token error, stopping listen loop.")
                            stopListening()
                        case .missingToken:
                            self.logger.error("Got missing token error, stopping listen loop.")
                            stopListening()
                        default:
                            break
                        }
                    }
                }.finally(on: DispatchQueue.global(qos: .utility)) {
                    self.isDoingListening = false
                    self.logger.info("Completed iteration of listen loop.")
                }
            }
        }
        timer.tolerance = 0.5
        RunLoop.current.add(timer, forMode: .common)
        RunLoop.current.run()
    }
    
    /**
    A `Promise` wrapper around Switrix's `SwitrixSyncClient().sync()` method.
     
     - Returns: A `Promise<SwitrixSyncResponse>`, in which the `SwitrixSyncResponse` is from the Commuto Interface Network test room.
     */
    private func syncPromise() -> Promise<SwitrixSyncResponse> {
        return Promise { seal in
            switrixClient.sync.sync { response in
                switch response {
                case .success(let switrixSyncResponse):
                    seal.fulfill(switrixSyncResponse)
                case .failure(let error):
                    seal.reject(error)
                }
            }
        }
    }
    
    #warning("TODO: would be nice to specify to parameter here")
    /**
    A `Promise` wrapper around Switrix's `SwitrixEventsClient().getEvents(...)` method. This method gets message events only from the Commuto Interface Network test room.
     
     - Parameters:
        - from: The batch token from which to start returning events.
        - limit: The maximun number of events to return. Note that this will likely be greater than the number of events actually returned, since Switrix filters out all events that are not text message events.
     
     - Returns: A  `Promise<SwitrixGetEventsResponse>`, in which the `SwitrixGetEventsResponse`, if successful, contains text message events from the Commuto Interface Network test room.
     */
    private func getMessagesPromise(from: String, limit: Int) -> Promise<SwitrixGetEventsResponse> {
        return Promise { seal in
            switrixClient.events.getEvents(roomId: "!WEuJJHaRpDvkbSveLu:matrix.org", from: from, direction: .forwards, limit: limit) { response in
                switch response {
                case .success(let switrixGetEventsResponse):
                    seal.fulfill(switrixGetEventsResponse)
                case .failure(let error):
                    seal.reject(error)
                }
            }
        }
    }
    
    /**
    Parses an array of `SwitrixClientEvent`s, filters out all non-message events, attempts to create message objects from the content bodies of message events, if present, and then passes any created offer-related message objects to `offerService`.
     
     - Parameter events: a list of `SwitrixClientEvent`s.
     */
    func parseEvents(_ events: [SwitrixClientEvent]) throws {
        let messageEvents = events.filter { event in
            return event.type == "m.room.message"
        }
        self.logger.notice("parseEvents: parsing \(messageEvents.count) m.room.message events")
        for event in messageEvents {
            // Handle unencrypted messages
            if let pka = parsePublicKeyAnnouncement(messageString: (event.content as? SwitrixMessageEventContent)?.body) {
                self.logger.notice("parseEvents: got Public Key Announcement message in event with Matrix event ID: \(event.eventId)")
                try offerService.handlePublicKeyAnnouncement(pka)
            } else {
                // If execution reaches this point, then we have already tried to get every possible unencrypted message from the event being handled. Therefore, we now try to get an encrypted message from the event: We attempt to turn the event content to data via UTF8 encoding, and then we attempt to create a JSON NSDictionary from that Data. If this is successful, we check for a recipient field, and attempt to create an interface ID from the contents of that field. Then we check keyManagerService to determine if we have a key pair with that interface ID. If we do, then we have determined that the event contains an encrypted message sent to us, and we attempt to parse it.
                guard let messageData = (event.content as? SwitrixMessageEventContent)?.body.data(using: String.Encoding.utf8) else {
                    // If we can't get Data from the event, then we stop handling it and move on
                    break
                }
                guard let message = try JSONSerialization.jsonObject(with: messageData) as? NSDictionary else {
                    // If we can't use JSON serialization to create an NSDictionary from the message data, we stop handling the event and move on
                    break
                }
                guard let recipientInterfaceIDString = message["recipient"] as? String, let recipientInterfaceID = Data(base64Encoded: recipientInterfaceIDString) else {
                    // If the message doesn't have a "recipient" field or if we can't create a recipient interface ID from the contents of the "recipient" field, then we stop handling it and move on
                    break
                }
                guard let recipientKeyPair = try keyManagerService.getKeyPair(interfaceId: recipientInterfaceID) else {
                    // If we don't have a key pair with the interface ID specified in the "recipient" field, then we don't have the private key necessary to decrypt the message, (meaning we aren't the intended recipient) so we stop handling it and move on
                    break
                }
                if let takerInformationMessage = try parseTakerInformationMessage(message: message, keyPair: recipientKeyPair) {
                    self.logger.notice("parseEvents: got Taker Information Message in event with Matrix event ID \(event.eventId)")
                    try swapService.handleTakerInformationMessage(takerInformationMessage)
                } else {
                    // If execution reaches this point, then we have already tried to get every possible encrypted message that doesn't require us to have the sender's public key. Therefore we check for a recipient field, and attempt to create an interface ID from the contents of that field, and then check keyManagerService to determine if we have a public key with that interface ID. If we do, then we continue attempting to parse the message. If we do not, we log a warning and break.
                    guard let senderInterfaceIDString = message["sender"] as? String, let senderInterfaceID = Data(base64Encoded: senderInterfaceIDString) else {
                        self.logger.warning("parseEvents: could not get sender interface ID from message in event with Matrix event ID \(event.eventId) sent to this interface")
                        break
                    }
                    guard let senderPublicKey = try keyManagerService.getPublicKey(interfaceId: senderInterfaceID) else {
                        self.logger.warning("parseEvents: could not find sender's public key for message sent to this interface in event with Matrix event ID \(event.eventId)")
                        break
                    }
                    if let makerInformationMessage = try parseMakerInformationMessage(message: message, keyPair: recipientKeyPair, publicKey: senderPublicKey) {
                        self.logger.notice("parseEvents: got Maker Information Message in event with Matrix event ID \(event.eventId)")
                        try swapService.handleMakerInformationMessage(makerInformationMessage, senderInterfaceID: senderInterfaceID, recipientInterfaceID: recipientInterfaceID)
                    }
                }
            }
        }
    
    }
    
    /**
     Sends the given message `String` in the Commuto Interface Network Test Room.
     
     - Parameter message: The `String` to send in the Commuto Interface Network Test Room.
     */
    func sendMessage(_ message: String) throws {
        _ = try Promise<SwitrixSendEventResponse> { seal in
            DispatchQueue.global(qos: .userInitiated).async { [self] in
                logger.notice("sendMessage: sending: \(message)")
                let eventContent = SwitrixMessageEventContent(body: message)
                switrixClient.rooms.sendEvent(roomId: "!WEuJJHaRpDvkbSveLu:matrix.org", eventType: "m.room.message", eventContent: eventContent) { [self] response in
                    switch response {
                    case .success(let sendMessageResponse):
                        logger.notice("sendMessage: success; ID: \(sendMessageResponse.eventId)")
                        seal.fulfill(sendMessageResponse)
                    case .failure(let error):
                        logger.error("sendMessage: got error: \(error.localizedDescription)")
                        seal.reject(error)
                    }
                }
            }
        }.wait()
        
    }
    
    /**
     Creates a [Public Key Announcement](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt) using the given offer ID and key pair and sends it using the `sendMessage` function.
     
     - Parameters:
        - offerID: The ID of the offer for which the Public Key Announcement is being made.
        - keyPair: The key pair containing the public key to be announced.
     */
    func announcePublicKey(offerID: UUID, keyPair: KeyPair) throws {
        logger.notice("announcePublicKey: creating for offer \(offerID.uuidString) and key pair with interface ID \(keyPair.interfaceId.base64EncodedString())")
        let announcement = try createPublicKeyAnnouncement(offerID: offerID, keyPair: keyPair)
        logger.notice("announcePublicKey: sending announcement for offer \(offerID.uuidString)")
        try sendMessage(announcement)
    }
    
    /**
     Creates a [Taker Information Message](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt) using the supplied maker's public key, taker's/user's key pair, swap ID and taker's payment details as an optional string, and sends it using the `sendMessage` function.
     
     - Parameters:
        - makerPublicKey: The public key of the swap maker, to whom information is being sent.
        - takerKeyPair: The taker's/user's key pair, which will be used to sign this message.
        - swapID: The ID of the swap for which information is being sent.
        - paymentDetails: The settlement method details being sent, as an optional string.
     */
    func sendTakerInformation(makerPublicKey: PublicKey, takerKeyPair: KeyPair, swapID: UUID, paymentDetails: String?) throws {
        logger.notice("sendTakerInformation: creating for \(swapID.uuidString)")
        let messageString = try createTakerInformationMessage(makerPublicKey: makerPublicKey, takerKeyPair: takerKeyPair, swapID: swapID, settlementMethodDetails: paymentDetails)
        logger.notice("sendTakerInformation: sending for \(swapID.uuidString)")
        try sendMessage(messageString)
    }
    
    /**
     Creates a [Maker Information Message](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt) using the supplied taker's public key, maker's/user's key pair, swap ID and maker's payment details, and sends it using the `sendMessage` function.
     
     - Parameters:
        - takerPublicKey: The public key of the swap taker, to whom information is being sent.
        - makerKeyPair: The maker's/user's key pair, which will be used to sign this message.
        - swapID: The ID of the swap for which information is being sent.
        - settlementMethodDetails: The settlement method details being sent, as an optional string.
     */
    func sendMakerInformation(takerPublicKey: PublicKey, makerKeyPair: KeyPair, swapID: UUID, settlementMethodDetails: String?) throws {
        logger.notice("sendMakerInformation: creating for \(swapID.uuidString)")
        let messageString = try createMakerInformationMessage(takerPublicKey: takerPublicKey, makerKeyPair: makerKeyPair, swapID: swapID, settlementMethodDetails: settlementMethodDetails)
        logger.notice("sendMakerInformation: sending for \(swapID.uuidString)")
        try sendMessage(messageString)
    }
    
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/p2p/P2PService.kt
// Suppress warning in initializer
@file:Suppress("LeakingThis")

package com.commuto.interfacemobile.android.p2p

import android.util.Log
import com.commuto.interfacemobile.android.key.KeyManagerService
import com.commuto.interfacemobile.android.key.keys.KeyPair
import com.commuto.interfacemobile.android.key.keys.PublicKey
import com.commuto.interfacemobile.android.offer.OfferService
import com.commuto.interfacemobile.android.p2p.create.createMakerInformationMessage
import com.commuto.interfacemobile.android.p2p.create.createTakerInformationMessage
import com.commuto.interfacemobile.android.p2p.parse.parseMakerInformationMessage
import com.commuto.interfacemobile.android.p2p.serializable.messages.SerializableEncryptedMessage
import com.commuto.interfacemobile.android.p2p.parse.parseTakerInformationMessage
import com.commuto.interfacemobile.android.swap.SwapService
import io.ktor.client.*
import io.ktor.client.plugins.*
import io.ktor.http.*
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import net.folivo.trixnity.clientserverapi.client.MatrixClientServerApiClient
import net.folivo.trixnity.clientserverapi.model.rooms.GetEvents
import net.folivo.trixnity.core.ErrorResponse
import net.folivo.trixnity.core.MatrixServerException
import net.folivo.trixnity.core.model.RoomId
import net.folivo.trixnity.core.model.events.Event
import net.folivo.trixnity.core.model.events.m.room.RoomMessageEventContent
import java.net.ConnectException
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The main Peer-to-Peer Service. It is responsible for listening to new messages in the Commuto
 * Interface Network Matrix room, parsing the messages, and then passing relevant messages to other
 * services as necessary.
 *
 * @constructor Creates a new [P2PService] with the specified [P2PExceptionNotifiable],
 * [OfferMessageNotifiable] and [MatrixClientServerApiClient].
 *
 * @property logTag The tag passed to [Log] calls.
 * @property exceptionHandler An object to which [P2PService] will pass errors when they occur.
 * @property offerService An object that implements [OfferMessageNotifiable], to which this will pass offer-related
 * messages.
 * @property swapService An object that implements [SwapMessageNotifiable], to which this will pass swap-related
 * messages.
 * @property mxClient A [MatrixClientServerApiClient] instance used to interact with a Matrix
 * Homeserver.
 * @property keyManagerService A [KeyManagerService] from which this gets key pairs when attempting to decrypt encrypted
 * messages.
 * @property lastNonEmptyBatchToken The token at the end of the last batch of non-empty Matrix
 * events that was parsed. (The value specified here is that from the beginning of the Commuto
 * Interface Network testing room.) This should be updated every time a new batch of events is
 * parsed.
 * @property listenJob The coroutine [Job] in which [P2PService] listens for and parses new batches
 * of Matrix events.
 * @property runLoop Boolean that indicates whether [listenLoop] should continue to execute its
 * loop.
 */
@Singleton
open class P2PService constructor(
    private val exceptionHandler: P2PExceptionNotifiable,
    private val offerService: OfferMessageNotifiable,
    private val swapService: SwapMessageNotifiable,
    private val mxClient: MatrixClientServerApiClient,
    private val keyManagerService: KeyManagerService,
) {

    @Inject
    constructor(
        exceptionHandler: P2PExceptionNotifiable,
        offerService: OfferMessageNotifiable,
        swapService: SwapMessageNotifiable,
        keyManagerService: KeyManagerService,
    ): this(
        exceptionHandler = exceptionHandler,
        offerService = offerService,
        swapService = swapService,
        mxClient = MatrixClientServerApiClient(
            baseUrl = Url("https://matrix.org"),
            httpClientFactory = {
                HttpClient(it).config {
                    install(HttpTimeout) {
                        socketTimeoutMillis = 60_000
                    }
                }
            }
        ).apply { accessToken.value = "syt_amltbXl0_TlbyjkdfhjbCVJNjOtOt_0GeYu6" },
        keyManagerService = keyManagerService
    )

    init {
        (offerService as? OfferService)?.setP2PService(this)
        (swapService as? SwapService)?.setP2PService(this)
    }

    private val logTag = "P2PService"

    private var lastNonEmptyBatchToken =
        "t1-2607497254_757284974_11441483_1402797642_1423439559_3319206_507472245_4060289024_0"

    /**
     * Used to update [lastNonEmptyBatchToken]. Eventually, this method will store [newToken] in
     * persistent storage.
     *
     * @param newToken The batch token at the end of the most recently parsed batch of Matrix
     * events, to be set as [lastNonEmptyBatchToken].
     */
    private fun updateLastNonEmptyBatchToken(newToken: String) {
        lastNonEmptyBatchToken = newToken
    }

    private var listenJob: Job = Job()

    private var runLoop = true

    /**
     * Launches a new coroutine [Job] in [GlobalScope], the global coroutine scope, runs
     * [listenLoop] in this new [Job], and stores a reference to it in [listenJob].
     */
    @OptIn(DelicateCoroutinesApi::class)
    fun listen() {
        Log.i(logTag, "Starting listen loop in global coroutine scope")
        listenJob = GlobalScope.launch {
            runLoop = true
            listenLoop()
        }
        Log.i(logTag, "Started listen loop in global coroutine scope")
    }

    /**
     * Sets [runLoop] to false to prevent the listen loop from being executed again and cancels
     * [listenJob].
     */
    fun stopListening() {
        Log.i(logTag, "Stopping listen loop and canceling listen job")
        runLoop = false
        listenJob.cancel()
        Log.i(logTag, "Stopped listen loop and canceled listen job")
    }

    /**
     * Executes the listening process in the current coroutine context as long as
     * [runLoop] is true.
     *
     * Listening Process:
     *
     * First, we sync with the Matrix Homeserver. Then we attempt to get all new Matrix events from
     * [lastNonEmptyBatchToken] to the present, with a limit of one trillion events. We then parse
     * these new events and then update the last non-empty batch token with the nextBatch token from
     * the sync response.
     *
     * If we encounter an [Exception], we pass it to [exceptionHandler]. Additionally, if the
     * exception is a [ConnectException], indicating that we are having problems communicating with
     * the homeserver, or if the exception is a [MatrixServerException], indicating that there is
     * something wrong with our Matrix API requests, then we stop listening.
     */
    suspend fun listenLoop() {
        while (runLoop) {
            try {
                // TODO: This works for now, but it may skip messages. replace it with something better.
                Log.i(logTag, "Beginning iteration of listen loop")
                val syncResponse = mxClient.sync.sync(timeout = 60_000).getOrThrow()
                Log.i(logTag, "Synced with Matrix homeserver, got nextBatchToken: ${syncResponse.nextBatch}")
                val eventsToParse = mxClient.rooms.getEvents(
                    roomId = RoomId(full = "!WEuJJHaRpDvkbSveLu:matrix.org"),
                    from = lastNonEmptyBatchToken,
                    dir = GetEvents.Direction.FORWARDS,
                    limit = 1_000_000_000_000L
                ).getOrThrow().chunk!!
                Log.i(logTag, "Got new messages from Matrix homeserver from lastNonEmptyBatchToken: " +
                        lastNonEmptyBatchToken)
                parseEvents(eventsToParse)
                Log.i(logTag, "Finished parsing new events from chunk of size ${eventsToParse.size}")
                updateLastNonEmptyBatchToken(syncResponse.nextBatch)
                Log.i(logTag, "Updated lastNonEmptyBatchToken with new token: ${syncResponse.nextBatch}")
            } catch (e: Exception) {
                Log.e(logTag, "Got an exception during listen loop, calling exception handler", e)
                exceptionHandler.handleP2PException(e)
                if (e is ConnectException ||
                    (e as? MatrixServerException)?.errorResponse is ErrorResponse.UnknownToken ||
                    (e as? MatrixServerException)?.errorResponse is ErrorResponse.MissingToken) {
                    Log.i(logTag, "Stopping listen loop for exception", e)
                    stopListening()
                }
            }
        }
    }

    /**
     * Parses a [List] of [Event.RoomEvent]s, filters out all non-text-message events, attempts to create message
     * objects from the content bodies of text message events, if present, and then passes any offer-related message
     * objects to [offerService] and any swap-related messages to [swapService].
     *
     * @param events A [List] of [Event.RoomEvent]s.
     */
    suspend fun parseEvents(events: List<Event.RoomEvent<*>>) {
        val textMessageEvents = events.filterIsInstance<Event.MessageEvent<*>>().filter {
            it.content is RoomMessageEventContent.TextMessageEventContent
        }
        Log.i(logTag, "parseEvents: parsing ${textMessageEvents.size} text message events")
        for (event in textMessageEvents) {
            // Handle unencrypted messages
            val pka = parsePublicKeyAnnouncement((event.content as RoomMessageEventContent.
            TextMessageEventContent).body)
            if (pka != null) {
                Log.i(logTag, "parseEvents: got Public Key Announcement message in event with Matrix event ID: " +
                        event.id.full)
                offerService.handlePublicKeyAnnouncement(pka)
            } else {
                /*
                If execution reaches this point, then we have already tried to get every possible unencrypted message
                from the event being handled. Therefore, we now try to get an encrypted message from the event: We
                attempt to cast the event content as text message event content, and then we attempt to create a
                SerializableEncryptedMessage from the text message event content body. Then we attempt to create an
                interface ID from the contents of the recipient field. Then we check keyManagerService to determine if
                we have a key pair with that interface ID. If we do, then we have determined that the event contains an
                encrypted message sent to us, and we attempt to parse it.
                 */
                val messageString = try {
                    (event.content as RoomMessageEventContent.TextMessageEventContent).body
                } catch (e: Exception) {
                    /*
                    If we can't can't get a message string from the room's content cast as TextMessageEventContent, then
                    we stop handling it and move on
                     */
                    break
                }
                val message = try {
                    Json.decodeFromString<SerializableEncryptedMessage>(messageString)
                } catch (e: Exception) {
                    /*
                    If we can't get a SerializableEncryptedMessage from the message, then we stop handling it and move
                    on
                     */
                    break
                }
                val decoder = Base64.getDecoder()
                val recipientInterfaceID = try {
                    decoder.decode(message.recipient)
                } catch (e: Exception) {
                    /*
                    If we can't create a recipient interface ID from the contents of the "recipient" field, then we stop
                    handling it and move on
                     */
                    break
                }
                /*
                If we don't have a key pair with the interface ID specified in the "recipient" field, then we don't
                have the private key necessary to decrypt the message, (meaning we aren't the intended recipient) so
                we stop handling it and move on
                */
                val recipientKeyPair = keyManagerService.getKeyPair(recipientInterfaceID)
                    ?: break
                val takerInformationMessage = parseTakerInformationMessage(
                    message = message,
                    keyPair = recipientKeyPair
                )
                if (takerInformationMessage != null) {
                    Log.i(logTag, "parseEvents: got Taker Information Message in event with Matrix event ID " +
                            event.id.full)
                    swapService.handleTakerInformationMessage(takerInformationMessage)
                    break
                }
                /*
                If execution reaches this point, then we have already tried to get every possible encrypted message that
                doesn't require us to have the sender's public key. Therefore we attempt to create an interface ID from
                the contents of that field, and then check keyManagerService to determine if we have a public key with
                that interface ID. If we do, then we continue attempting to parse the message. If we do not, we log a
                warning and break.
                 */
                val senderInterfaceID = try {
                    decoder.decode(message.sender)
                } catch (e: Exception) {
                    break
                }
                val senderPublicKey = keyManagerService.getPublicKey(senderInterfaceID)
                    ?: break
                val makerInformationMessage = parseMakerInformationMessage(
                    message = message,
                    keyPair = recipientKeyPair,
                    publicKey = senderPublicKey
                )
                if (makerInformationMessage != null) {
                    Log.i(logTag, "parseEvents: got Maker Information Message in event with Matrix event ID " +
                            event.id.full)
                    swapService.handleMakerInformationMessage(
                        message = makerInformationMessage,
                        senderInterfaceID = senderInterfaceID,
                        recipientInterfaceID = recipientInterfaceID
                    )
                }
            }
        }
    }

    /**
     * Sends the given message [String] in the Commuto Interface Network Test Room.
     *
     * @param message The [String] to send in the Commuto Interface Network Test Room.
     */
    open suspend fun sendMessage(message: String) {
        Log.i(logTag, "sendMessage: sending $message")
        val result = mxClient.rooms.sendMessageEvent(
            roomId = RoomId("!WEuJJHaRpDvkbSveLu:matrix.org"),
            eventContent = RoomMessageEventContent.TextMessageEventContent(message),
        ).getOrElse {
            Log.e(logTag, "sendMessage: got exception", it)
            throw it
        }
        Log.i(logTag, "sendMessage: success; ID: $result")
    }

    /**
     * Creates a
     * [Public Key Announcement](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt)
     * using the given offer ID and key pair and sends it using the [sendMessage] function.
     *
     * @param offerID The ID of the offer for which the Public Key Announcement is being made.
     * @param keyPair The key pair containing the public key to be announced.
     */
    open suspend fun announcePublicKey(offerID: UUID, keyPair: KeyPair) {
        val encoder = Base64.getEncoder()
        Log.i(logTag, "announcePublicKey: creating for offer $offerID and key pair with interface ID " +
                encoder.encodeToString(keyPair.interfaceId)
        )
        val announcement = createPublicKeyAnnouncement(offerID = offerID, keyPair = keyPair)
        Log.i(logTag, "announcePublicKey: sending announcement for offer $offerID")
        sendMessage(announcement)
    }

    /**
     * Creates a
     * [Taker Information Message](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt)
     * using the supplied maker's public key, taker's/user's key pair, swap ID and taker's payment details, and sends it
     * using the [sendMessage] function.
     *
     * @param makerPublicKey The public key of the swap maker, to whom information is being sent.
     * @param takerKeyPair The taker's/user's key pair, which will be used to sign this message.
     * @param swapID The ID of the swap for which information is being sent.
     * @param settlementMethodDetails The settlement method details being sent.
     */
    open suspend fun sendTakerInformation(
        makerPublicKey: PublicKey,
        takerKeyPair: KeyPair,
        swapID: UUID,
        settlementMethodDetails: String,
    ) {
        Log.i(logTag, "sendTakerInformation: creating for $swapID")
        val messageString = createTakerInformationMessage(
            makerPublicKey = makerPublicKey,
            takerKeyPair = takerKeyPair,
            swapID = swapID,
            settlementMethodDetails = settlementMethodDetails
        )
        Log.i(logTag, "sendTakerInformation: sending for $swapID")
        sendMessage(messageString)
    }

    /**
     * Creates a
     * [Maker Information Message](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt)
     * using the supplied taker's public key, maker's/user's key pair, swap ID and maker's payment details, and sends it
     * using the [sendMessage] function.
     *
     * @param takerPublicKey The public key of the swap taker, to whom information is being sent.
     * @param makerKeyPair The maker's/user's key pair, which will be used to sign this message.
     * @param swapID The ID of the swap for which information is being sent.
     * @param settlementMethodDetails The settlement method details being sent.
     */
    open suspend fun sendMakerInformation(
        takerPublicKey: PublicKey,
        makerKeyPair: KeyPair,
        swapID: UUID,
        settlementMethodDetails: String,
    ) {
        Log.i(logTag, "sendMakerInformation: creating for $swapID")
        val messageString = createMakerInformationMessage(
            takerPublicKey = takerPublicKey,
            makerKeyPair = makerKeyPair,
            swapID = swapID,
            settlementMethodDetails = settlementMethodDetails
        )
        Log.i(logTag, "sendMakerInformation: sending for $swapID")
        sendMessage(messageString)
    }

}